# 본문의 글은 인프런 앨런 Swift문법 마스터 스쿨(온라인) 부트캠프의 강의 영상을 본후 공부한 내용을 정리하여 기록한것입니다.
https://inf.run/YyoR

```swift


//런타임 오류
//1. 여러가지 다양한 이유로 앱이 실행되는 동안 발생 => 앱이 꺼짐
//2. 미리발생가능한 에러를파악, 올바르게 처리하면 앱의 꺼짐을 막을수 있음.

//런타임 에러처리가 왜 필요할까?
//예기치 못하는 상황이 발생할수 있따. 서버에서 데이터를 못받아온다든가 하는 경우
//그런경우 그냥 앱이 꺼지는것 보단 "문제가발생했습니다. 잠시뒤에 접속해주세요" 라고하는게 좋다.
//프로세스중에서 에러발생가능성을 처리해두면 앱이무작정꺼지는것을 예방할수있음
//에러는 일반적으로 함수의 처리과정에서 일어나고 에러가 날수있는 함수를 정의할수 있음
//에러발생가능함수 ===> 함수실행시 조금 다르게 처리해야함


//일반 적인 함수
func doSomething(num: Int) -> Bool {
    if num > 7 {
        return true
    } else {
        return false
    }
}


doSomething(num: 8)


//열거형 에러
enum SomeError: Error { //에러를정의를 하려면 에러 프로토콜을 채택해야함
    case aError
    case bError
    case cError
}

//에러를 던지는 함수
func doSomething2(num: Int) throws -> Bool { //throws 키워드
    if num > 7 {
        return true
    } else {
        
        if num < 0 {
            throw SomeError.aError  //리턴하고비슷한 throw, 열거형 에러
        }
        
        return false
    }
}
//정상적인경우는 true나 false를 리턴하고 비정상적인상황인 경우는 에러를 던진다...






//에러처리과정의 3단계 ====================================

//1. 에러 정의 (어떤 에러가 발생할지 경우를 미리 정의)

enum HeightError: Error { //에러프로토콜을 채택 꼭
    case maxHeight
    case minHeight
}

//2. 에러가발생할수있는 함수에 대한 저으이

func checkingHeight(height: Int) throws -> Bool {
    if height > 190 {
        throw HeightError.maxHeight
    } else if height < 130 {
        throw HeightError.minHeight
    } else {
        if height > 160 {
            return true
        } else {
            return false
        }
    }
}


//3. 에러가 발생할수있는 함수의 처리 (함수의 실행) ===> 기존과 조금 다르다 (try , do-catch 문으로 처리)
//반드시 에러발생가능의 함수실행은 do블럭안에서 써야하고 try라는 키워드를 붙혀야함
//리턴값이 있는경우는 변수에 담을수도 있음

do { //정상적인 경우의 처리 상황
    let result = try checkingHeight(height: 110)
    if result {
        print("참")
    } else {
        print("거짓")
    }
} catch { //실제로 위에코드가 에러를 던지는 경우의 처리를 하는 곳
    print("에러입니다~~~~")
}

// () throws -> () 에러를 발생시킬수있는 함수의 타입이구나~ 하고 알아야한다.

// [1,2,3,4,5].map(<#T##transform: (Int) throws -> T##(Int) throws -> T#>) <= 이형태가 바로 에러를 던지는 함수였떤것 (map고차함수)





enum HeightError: Error { //에러프로토콜을 채택 꼭
    case maxHeight
    case minHeight
}

func checkingHeight(height: Int) throws -> Bool {
    if height > 190 {
        throw HeightError.maxHeight
    } else if height < 130 {
        throw HeightError.minHeight
    } else {
        if height > 160 {
            return true
        } else {
            return false
        }
    }
}


//에러처리방법 try/try?/try! ========================

//1. 정식적인 처리 방법
do {
    let isChecked = try checkingHeight(height: 200)
    
    if isChecked {
        print("청룡열차 쌉가능")
    } else {
        print("후룸라이드도 쌉가능")
    }
} catch {
    print("키가 너무크거나 작아 놀이기구 못타요~~")
}

//2. 옵셔널 try? ==> 옵셔널타입으로 리턴
//정상적인경우 => 정상리턴타입으로 리턴
//에러가발생하면 => nil 리턴

let isChecked = try? checkingHeight(height: 200)
if isChecked == nil {
    print("키가 너무크거나 작아 놀이기구못타요")
}
//당연히 벗겨서 사용해야함



//3. try! => 에러가날수없는경우에만 사용 가능

let isChecked2 = try! checkingHeight(height: 150)

//에러가 발생할수없다고 확신이 있는경우에만 사용해야함





//catch 블럭 처리법 ======================

//catch블럭은 do블럭에서 발생한 에러만을 처리하는블럭
// 모든 에러를 반드시 처리해야만 함 (글로벌 스코프에서는 모든에러를 처리하지 않아도 컴파일 에러 발생하지않음)


do {
    let isChecked = try checkingHeight(height: 80)
    print("놀이기구 탈수있는 키 입니다: \(isChecked)")
} catch HeightError.maxHeight {  //where 절 추가로 매칭시킬 에러패턴에 조건을 추가할수도 있음(자주사용x)
    print("키가 너무커서 놀이기구 못타요")
} catch HeightError.minHeight { //생략가능 그냥 catch만 남기면 스위치문의 디폴트문 처럼 사용 가능
    print("키가 너무작아서 놀이이구 못타요")
}



//catch 패턴없이도 처리 가능

do {
    let isChecked = try checkingHeight(height: 200)
    print("놀이기구 탈수있는 키 입니다: \(isChecked)")
} catch {
    print(error.localizedDescription)  //error 라는 Error 프로토콜을 담고있는 상수를 제공해줌(자동으로)
        //.localizedDescription 는 지역언어로 출력해줌
    if let error = error as? HeightError { //실제 우리가 정의한 열거형으로 타입캐스팅한것
    //error 라는 상수는 Error프로토콜자체를 담고있는녀석, heightError 라는녀석은 에러프로토콜을 채택한 녀석이라 다운캐스팅이 가능하다.
        switch error { //스위치문으로 구체적으로나눈것
        case .maxHeight:
            print("키가 너무커서 놀이기구 못타요")
        case .minHeight:
            print("키가 너무작아서 놀이기구 못타요")
        }
    }
}




//swift 5.3 이후 ===========================

do {
    let isChecked = try checkingHeight(height: 100)
    print("놀이기구 탈수있는 키 입니다: \(isChecked)")
} catch HeightError.maxHeight, HeightError.minHeight { //,를 찍어서 케이스를 나열하수도 있음
    print("놀이기구 못탑니데이")
}

```